src/block/inventory/EnchantInventory.php

```diff
@@ -25,16 +25,78 @@
 
 use pocketmine\inventory\SimpleInventory;
 use pocketmine\inventory\TemporaryInventory;
+use pocketmine\item\enchantment\EnchantmentHelper as Helper;
+use pocketmine\item\enchantment\EnchantmentOption;
+use pocketmine\item\Item;
+use pocketmine\item\ItemTypeIds;
+use pocketmine\item\VanillaItems;
 use pocketmine\world\Position;
+use function count;
 
 class EnchantInventory extends SimpleInventory implements BlockInventory, TemporaryInventory{
 	use BlockInventoryTrait;
 
 	public const SLOT_INPUT = 0;
 	public const SLOT_LAPIS = 1;
 
+	/** @var EnchantmentOption[] $options */
+	private array $options = [];
+
+	/** @var Item[] $outputs */
+	private array $outputs = [];
+
 	public function __construct(Position $holder){
 		$this->holder = $holder;
 		parent::__construct(2);
 	}
+
+	protected function onSlotChange(int $index, Item $before) : void{
+		if($index == self::SLOT_INPUT){
+			foreach($this->viewers as $viewer){
+				$this->options = Helper::getEnchantOptions($this->holder, $this->getItem(self::SLOT_INPUT), $viewer->getXpSeed());
+				$this->outputs = [];
+				if (count($this->options) !== 0) {
+					$viewer->getNetworkSession()->sendEnchantOptions($this->options);
+				}
+			}
+		}
+
+		parent::onSlotChange($index, $before);
+	}
+
+	public function getInput() : Item{
+		return $this->getItem(self::SLOT_INPUT);
+	}
+
+	public function getLapis() : Item{
+		return $this->getItem(self::SLOT_LAPIS);
+	}
+
+	public function getOutput(int $optionId) : ?Item{
+		$this->prepareOutput($optionId);
+		return $this->outputs[$optionId] ?? null;
+	}
+
+	private function prepareOutput(int $optionId) : void{
+		if(isset($this->outputs[$optionId])){
+			return;
+		}
+
+		$option = $this->options[$optionId] ?? null;
+		if($option === null){
+			// Failed to find an enchantment option with the passed network id
+			return;
+		}
+
+		$outputItem = $this->getItem(self::SLOT_INPUT);
+		if($outputItem->getTypeId() === ItemTypeIds::BOOK){
+			$outputItem = VanillaItems::ENCHANTED_BOOK();
+		}
+
+		foreach($option->getEnchantments() as $enchantment){
+			$outputItem->addEnchantment($enchantment);
+		}
+
+		$this->outputs[$optionId] = $outputItem;
```

Comment:
Why are you caching the output, it's only used once and then resets after transaction completion, isn't it?

Reply:
That's right, I had originally planned to check if the outputItem in EnchantTransaction is correct, but then realized it is not needed) I'll remove caching